src/manifold/src/manifold.cpp

```diff
@@ -138,6 +138,56 @@ Mesh Manifold::GetMesh() const {
   return result;
 }
 
+size_t Manifold::GetMeshDirect(int mode) const {
+    const Impl& impl = *GetCsgLeafNode().GetImpl();
+
+    int nv = NumVert();
+    int nt = NumTri();
+
+    int recs = 6;
+    switch (mode) {
+        // return just the header with pointers to data in wasm memory
+        case 0: break;
+        // return flattened index array after header
+        case 1: recs += (nt * 3); break;
+        // return unrolled and flattened vertex array after header
+        case 2: recs += (nt * 9); break;
+    }
+
+    int ms = sizeof(size_t) * recs;
+    size_t *x = (size_t *)malloc(ms);
+    float *f = (float *)x;
+
+    int rp = 0;
+
+    // fixed header
+    x[rp++] = (size_t)impl.vertPos_.begin();
+    x[rp++] = (size_t)impl.vertPos_.end();
+    x[rp++] = nv;
+    x[rp++] = (size_t)impl.halfedge_.begin();
+    x[rp++] = (size_t)impl.halfedge_.end();
+    x[rp++] = nt;
+
+    if (mode == 1) {
+        for (int i = 0; i < nt * 3; i++) {
+            x[rp++] = impl.halfedge_[i].startVert;
+        }
+    } else if (mode == 2) {
+        for (int i = 0; i < nt * 3; i++) {
+            const glm::vec3 v = impl.vertPos_[ impl.halfedge_[i].startVert ];
+            f[rp++] = v.x;
+            f[rp++] = v.y;
+            f[rp++] = v.z;
```

Comment:
Since you're getting 100x speed increase, rather than 2x, it makes me think this is less about having one copy rather than two, and more about using a fast C++-style copy rather than a slow JS array-of-arrays copy. If that's the case, then it would be nice to avoid all this old-school C-pointer stuff and new API, and just use two fast copies. Just trying to balance maintainability against performance.

Reply:
this is mode 2 which produces a flattened vertex array in WASM. I threw that in because when you want the vertices in this format, it is a lot faster than doing it in JS. mode 1, based on benchmarking, is the one I most care about. it still has old school pointer stuff, though. I think it's unavoidable when you just want to hand JS the location of the arrays on the heap for fast copy-out. Emscripten is hiding stuff like this somewhere in its bindings so that the rest of the code looks neat.